src/include/migraphx/op/slice.hpp

```diff
@@ -90,80 +190,141 @@ struct slice
+                           {
+                               index = dim;
+                           }
+                           return index;
+                       });
+        return indices;
+    }
 
-        std::vector<std::size_t> new_lens = old_lens;
-        for(std::size_t i = 0; i < axes.size(); i++)
-        {
-            auto axis            = axes[i];
-            size_t sliced_length = ends[i] - starts[i];
-            // A Numpy indexing convention: a slice size larger than the actual dimension
-            // is legal and the "ends" value is clipped to the axis size
-            new_lens[axis] = std::min(new_lens[axis], sliced_length);
-            if(input_shape.dynamic())
+    template <class Inds>
+    std::unordered_map<std::string, Inds>
+    normalize_inputs(shape input_shape, Inds input_starts, Inds input_ends) const
+    {
+        return {{"input_starts", normalize_indices(input_shape, input_starts, this->axes)},
+                {"input_ends", normalize_indices(input_shape, input_ends, this->axes)}};
+    }
+
+    /**
+     * Three input version of the normalize_inputs.
+     * This one also checks that the input_axes are valid.
+     */
+    template <class Inds>
+    std::unordered_map<std::string, Inds>
+    normalize_inputs(shape input_shape, Inds input_starts, Inds input_ends, Inds input_axes) const
+    {
+        // normalize the axes
+        auto shape_ndim = input_shape.ndim();
+        std::transform(input_axes.begin(), input_axes.end(), input_axes.begin(), [&](auto axis) {
+            if(axis < 0)
             {
-                // TODO: when non-fixed shape slicing is allowed, this will be different than
-                // sliced_length, making use of TBD start/end values.
-                std::size_t sliced_min_length = ends[i] - starts[i];
-                // if the slice size is smaller than maxes but larger than mins
-                new_mins[axis] = std::min(sliced_min_length, new_mins[axis]);
+                axis += shape_ndim;
             }
-        }
-        if(input_shape.dynamic())
+            if(axis < 0 or axis >= shape_ndim)
+            {
+                MIGRAPHX_THROW("SLICE: entry of input_axes out of bounds");
+            }
+            return axis;
+        });
+        return {{"input_starts", normalize_indices(input_shape, input_starts, input_axes)},
+                {"input_ends", normalize_indices(input_shape, input_ends, input_axes)},
+                {"input_axes", input_axes}};
+    }
+
+    argument compute(const shape&, std::vector<argument> args) const
+    {
+        auto input       = args[0];
+        auto input_shape = input.get_shape();
+        std::size_t offset;
+        switch(args.size())
         {
-            return shape{t, new_mins, new_lens, {}};
+        case 1: {
+            offset = compute_offset(input_shape);
+            return {normalize_compute_shape({input_shape}), [=] { return input.data() + offset; }};
```

Comment:
Why doesnt this use the `output_shape` that got passed in?

Reply:
It can if we add a conditional on if the shape is dynamic to recompute